### PR TITLE
nixos/wrappers: make mount have the +s bit.

### PR DESCRIPTION
**Pull request by doronbehar**
See
https://discourse.nixos.org/t/how-to-make-a-derivations-executables-have-the-s-permission/8555
and:
https://www.linuxquestions.org/questions/slackware-14/must-be-superuser-to-use-mount-fstab-is-correct-however-144932/

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by khumba** [comment]
Hello, there was some previous discussion on this at #9848.  I still support doing this, it would be useful, maybe behind a option for those that don't want setuid, and/or a warning if `user` or `users` is used and mount isn't setuid.  As far as I know, there was no follow-up from that issue.

**Comment by doronbehar** [comment]
Wow this has been talked over a lot! I still don't understand the security concern against the mere assertion you wanted to suggest in https://github.com/NixOS/nixpkgs/pull/9998 . Do you know how much time I wasted on trying to find this out?

For the protocol, I can't use udisks as I'm not interested in mounting a device. I want to `mount --bind` a local directory.

@edolstra you are the boss. I'd like to save future users at least the time I spent debugging this - If you are against even a mere assertion such as in https://github.com/NixOS/nixpkgs/pull/9998/files :

```nix
          any (fs: [] != intersectLists (splitString "," fs.options)
                                        [ "owner" "user" "users" ])
              fileSystems
          -> all (x: elem x config.security.setuidPrograms)
                 [ "mount" "umount" ];
        message = ''
          The fstab "owner", "user", and "users" options require "mount" and "umount"
          to be setuid.  Please add the following to your system configuration:
              security.setuidPrograms = [ "mount" "umount" ];
        '';
```

How about only mentioning this in the `options` description? Example:

```patch
diff --git i/nixos/modules/tasks/filesystems.nix w/nixos/modules/tasks/filesystems.nix
index 0ade74b957a..796d8e8220f 100644
--- i/nixos/modules/tasks/filesystems.nix
+++ w/nixos/modules/tasks/filesystems.nix
@@ -52,9 +52,16 @@ let
       options = mkOption {
         default = [ "defaults" ];
         example = [ "data=journal" ];
-        description = "Options used to mount the file system.";
-        type = types.listOf nonEmptyStr;
-      };
+        description = ''
+          Options used to mount the file system - the . Note that using either of the
+          options <literal>users</literal>, <literal>user</literal>,
+          <literal>owner</literal> will require you to make `mount` have the +s
+          permission, which is a known security risk. See <link
+          xlink:href="https://www.linuxquestions.org/questions/slackware-14/must-be-superuser-to-use-mount-fstab-is-correct-however-144932/"
+          >this thread</link> and discussion at <link
+          xlink:href="https://github.com/NixOS/nixpkgs/issues/9848">#9848</link>
+        '';
+        type = types.listOf nonEmptyStr; };
 
     };
```

**Review by flokli — APPROVED** [review]
`udisks` is proposed is an alternative tool to mount things in https://github.com/NixOS/nixpkgs/issues/9848, but it's more tailored towards allowing users to mount block devices ad-hoc (which aren't in `/etc/fstab` anyway).

There's still usecases for the `users` option in `/etc/fstab` (which requires the suid bit via the wrapper) - while it's not particularily nice, it's better than silently keeping things broken, so :+1: for me on this PR.

**Comment by doronbehar** [comment]
Thanks @flokli. The last push adds `umount` in the same manner.

**Comment by flokli** [comment]
Thanks! Can you extend one of the VM tests to check this?

**Comment by doronbehar** [comment]
Sure. I tested both the use case of expecting a `mount` denial when `users` is not specified and both my original use case.

File: nixos/tests/misc.nix [hunk 1]
```diff
@@ -112,6 +124,22 @@ import ./make-test-python.nix ({ pkgs, ...} : rec {
...
 
+      with subtest("Whether mounting by a user is possible (PR 95444)"):
+          machine.succeed("mkdir -p /user-mount/source")
+          machine.succeed("touch /user-mount/source/file")
+          machine.succeed("chmod -R a+Xr /user-mount/source")
+          machine.succeed("mkdir /user-mount/point")
+          machine.succeed("chown alice:users /user-mount/point")
+          machine.succeed("su - alice -c 'mount /user-mount/point'")
+          machine.succeed("su - alice -c 'ls /user-mount/point/file'")
+      with subtest("Whether mounting by a user is denied if user is not specified"):
```
Comment (flokli) [inline nixos/tests/misc.nix:135]:
```suggestion
      with subtest("Whether mounting by a user is denied without the `user` option in  fstab"):
```

File: nixos/tests/misc.nix [hunk 2]
```diff
@@ -112,6 +124,22 @@ import ./make-test-python.nix ({ pkgs, ...} : rec {
           machine.succeed("touch /tmp2/x")
           machine.succeed("grep '/tmp2 tmpfs' /proc/mounts")
 
+      with subtest("Whether mounting by a user is possible (PR 95444)"):
```
Comment (flokli) [inline nixos/tests/misc.nix:127]:
```suggestion
      with subtest("Whether mounting by a user is possible with the `user` option in fstab (#95444)"):
```

**Comment by flokli** [comment]
`black` isn't happy, and asks to reflow. Can you do that, and also squash that (and the other suggestions) into d03da650d79b4b2b17768bd8f6aa021a696a35d0?

**Comment by doronbehar** [comment]
> black isn't happy, and asks to reflow.

Who is black?

**Comment by flokli** [comment]
The code formatter - try running the test after the latest version of this PR.

**Comment by doronbehar** [comment]
Oh I see - it's due to the now longer subtest descriptions. Now it's fixed.

**Pull request merged**